lib/modules/manager/gradle/parser/common.ts

```diff
@@ -0,0 +1,101 @@
+import type { lexer } from 'good-enough-parser';
+import type { Ctx, NonEmptyArray, PackageVariables } from '../types';
+
+export const REGISTRY_URLS = {
+  google: 'https://dl.google.com/android/maven2/',
+  gradlePluginPortal: 'https://plugins.gradle.org/m2/',
+  jcenter: 'https://jcenter.bintray.com/',
+  mavenCentral: 'https://repo.maven.apache.org/maven2',
+};
+
+export const ANNOYING_METHODS: ReadonlySet<string> = new Set([
+  'createXmlValueRemover',
+  'events',
+  'args',
+  'arrayOf',
+  'listOf',
+  'mutableListOf',
+  'setOf',
+  'mutableSetOf',
+  'stages', // https://github.com/ajoberstar/reckon,
+  'mapScalar', // https://github.com/apollographql/apollo-kotlin
+]);
+
+export function storeVarToken(ctx: Ctx, node: lexer.Token): Ctx {
+  ctx.varTokens.push(node);
+  return ctx;
+}
+
+export function storeInTokenMap(ctx: Ctx, tokenMapKey: string): Ctx {
+  ctx.tokenMap[tokenMapKey] = ctx.varTokens;
+  ctx.varTokens = [];
+
+  return ctx;
+}
+
+export function loadFromTokenMap(
+  ctx: Ctx,
+  tokenMapKey: string
+): NonEmptyArray<lexer.Token> {
+  const tokens = ctx.tokenMap[tokenMapKey];
+  if (!tokens) {
+    throw new Error(`Expected token ${tokenMapKey} not found`);
+  }
+
+  return tokens as NonEmptyArray<lexer.Token>;
+}
+
+export function cleanupTempVars(ctx: Ctx): Ctx {
+  ctx.tokenMap = {};
+  ctx.varTokens = [];
+
+  return ctx;
+}
+
+export function stripReservedPrefixFromKeyTokens(ctx: Ctx): Ctx {
+  const unwantedPrefixes = ['ext', 'extra', 'project', 'rootProject'];
+  while (
+    ctx.varTokens.length > 1 && // ensures there will be always at least one token
+    ctx.varTokens[0] &&
+    unwantedPrefixes.includes(ctx.varTokens[0].value)
+  ) {
+    ctx.varTokens.shift();
+  }
+
+  return ctx;
+}
+
+export function coalesceVariable(ctx: Ctx): Ctx {
+  if (ctx.varTokens.length > 1) {
+    ctx.varTokens[0]!.value = ctx.varTokens
+      .map((token) => token.value)
+      .join('.');
+    ctx.varTokens.length = 1;
+  }
+
+  return ctx;
+}
+
+export function interpolateString(
+  childTokens: lexer.Token[],
+  variables: PackageVariables
+): string | null {
+  const resolvedSubstrings: string[] = [];
+  for (const childToken of childTokens) {
```

Comment:
you are running on child tokens, and if one has no data you return null ignoring the rest of the loop, is that intentional?

Reply:
Yes, it is. The method only returns != null if interpolation worked for all substituted segments